src/main/java/org/jvnet/hudson/test/WarExploder.java

```diff
@@ -147,30 +152,54 @@ private static File explode() throws Exception {
-            new FilePath(explodeDir).deleteRecursive();
-            new FilePath(war).unzip(new FilePath(explodeDir));
-            if(!explodeDir.exists())    // this is supposed to be impossible, but I'm investigating HUDSON-2605
-                throw new IOException("Failed to explode "+war);
-            new FileOutputStream(timestamp).close();
-            timestamp.setLastModified(war.lastModified());
-            new File(explodeDir + ".exploding").delete();
-        } else {
-            LOGGER.log(Level.INFO, "Picking up existing exploded jenkins.war at {0}", explodeDir.getAbsolutePath());
-        }
 
+        // multiple surefire forks can be running in parallel (which are different processes)
+        // so we can not use synchronisation here.
+        Path lock = new File(explodeDir + ".lock").toPath();
+        // it is not the presence of the lock file that prevents reading / writing (as that can not be made reliable
+        // but the lock we subsequently obtain on the file.
+        try (FileChannel lockChannel = FileChannel.open(lock, StandardOpenOption.CREATE, StandardOpenOption.READ, StandardOpenOption.WRITE);
+                FileLock fl = getLockForChannel(lockChannel)) {
+            File timestamp = new File(explodeDir,".timestamp");
+            if(!timestamp.exists() || (timestamp.lastModified() != war.lastModified())) {
+                LOGGER.log(Level.INFO, "Exploding {0} into {1}", new Object[] {war, explodeDir});
+                new FilePath(explodeDir).deleteRecursive();
+                new FilePath(war).unzip(new FilePath(explodeDir));
+                timestamp.createNewFile();
+                timestamp.setLastModified(war.lastModified());
+            } else {
+                LOGGER.log(Level.INFO, "Picking up existing exploded jenkins.war at {0}", explodeDir.getAbsolutePath());
+            }
+        }
         return explodeDir;
     }
 
+    private static FileLock getLockForChannel(FileChannel channel) throws IOException, InterruptedException {
+        FileLock lock = null;
+        int iteration = 0;
+        while (lock == null) {
+            try {
+                lock = channel.tryLock();
+            } catch (OverlappingFileLockException ignored) {
+                // should only occur when we have multiple threads in this JVM attempting to lock this file
+                // by default surefire and junit use JVM per fork - but gradle and other testing frameworks may differ
+                // so be defensive and treat this specific exception as a failure to obtain the lock rather than a 
+                // generic failure
+            }
+            if (lock == null) {
+                if (++iteration % 50 == 0) {
+                    // only log every 5 seconds.
+                    LOGGER.log(Level.INFO, "Waiting for a different JVM or thread to finish unpacking the war");
+                }
+                Thread.sleep(100);
+            }
+        }
+        return lock;
```

Comment:
Arguably clearer:
```suggestion
        int iteration = 0;
        while (true) {
            try {
                FileLock lock = channel.tryLock();
                if (lock != null) {
                    return lock;
                }
            } catch (OverlappingFileLockException ignored) {
                // should only occur when we have multiple threads in this JVM attempting to lock this file
                // by default surefire and junit use JVM per fork - but gradle and other testing frameworks may differ
                // so be defensive and treat this specific exception as a failure to obtain the lock rather than a 
                // generic failure
            }
            if (++iteration % 50 == 0) {
                // only log every 5 seconds.
                LOGGER.log(Level.INFO, "Waiting for a different JVM or thread to finish unpacking the war");
            }
            Thread.sleep(100);
        }
```